src/libos/crates/async-rt/src/sync/mutex/inner.rs

```diff
@@ -0,0 +1,221 @@
+struct AtomicLockStatus(AtomicI32);
+
+// The status code keeps the same as the implementation in musl.
+#[derive(Debug, Copy, Clone)]
+#[repr(i32)]
+enum LockStatus {
+    Free = 0,
+    Locked = Errno::EBUSY as i32,
+    LockedWithWaiters = -2147483632, // 0x8000_0010 = (EBUSY | 0x8000_0000)
+}
+
+impl MutexInner {
+    pub(super) fn new() -> MutexInner {
+        MutexInner {
+            status: AtomicLockStatus::init(),
+            waiters: AtomicI32::new(0),
+            waiter_queue: Mutex::new(WaiterQueue::new()),
+        }
+    }
+
+    pub(super) fn try_lock(&self) -> Result<()> {
+        if self.status.try_set_lock().is_ok() {
+            Ok(())
+        } else {
+            Err(errno!(EBUSY, "the lock is held by other threads"))
+        }
+    }
+
+    pub(super) async fn lock(&self) -> Result<()> {
+        if let Ok(_) = self.try_lock() {
+            return Ok(());
+        }
+
+        // Spin for a short while if no one is waiting but the lock is held.
+        let mut spins = SPIN_COUNT;
+        while spins != 0
+            && self.status.is_locked()
+            // Can't reorder here. `Relaxed` is enough.
+            && self.waiters.load(Ordering::Relaxed) == 0
+        {
+            hint::spin_loop();
+            spins -= 1;
+        }
+
+        loop {
+            if let Ok(_) = self.try_lock() {
+                return Ok(());
+            }
+
+            if self.status.is_free() {
+                continue;
+            }
+
+            // In try_set_lock_with_waiters, `AcqRel` will make sure this increment happens before. Thus, `Relaxed` can be used here.
+            self.waiters.fetch_add(1, Ordering::Relaxed);
+
+            // Ignore the result here. If the state transition fails, the next wait will not block.
+            self.status
+                .try_set_lock_with_waiters()
+                .map_err(|e| errno!(e.errno(), "failed to set lock status"))
+                .ok();
+
+            let ret = self.wait().await;
+            // Use `Acquire` to make sure `ret` is already set.
+            self.waiters.fetch_sub(1, Ordering::Acquire);
+
+            if let Err(error) = &ret {
+                match error.errno() {
+                    Errno::EINTR => continue,
+                    _ => {
+                        return_errno!(error.errno(), "futex wait error")
+                    }
+                }
+            }
+        }
+    }
+
+    pub(super) fn unlock(&self) -> Result<()> {
+        // `get_current_status_and_set_free` will make sure this happens before. Thus, `Relaxed` can be used here.
+        let waiters = self.waiters.load(Ordering::Relaxed);
+        let previous_status = self.status.get_current_status_and_set_free();
+
+        // Also check previous status in case the waiter number overflows.
+        if waiters > 0 || previous_status.is_locked_with_waiters() {
+            self.wake_one_waiter();
+        }
+
+        Ok(())
+    }
+
+    async fn wait(&self) -> Result<()> {
```

Comment:
My understanding is that there is a single caller of this method. And this caller does not care about the result of this method: it always retries regardless of the outcome. Since the caller and callee are all written by you, why bother return a useless result?